### PR TITLE
#662 required attributes warning for pivoted shape

### PR DESCRIPTION
**Pull request by GDonRanasinghe**
Satisfied #662

### CODE REVIEW

File: src/csvcubed/writers/qbwriter.py [hunk 1]
```diff
@@ -540,19 +540,19 @@ def _get_primary_key_columns(self) -> List[str]:
...
-                # since we permit missing observation values where an `obsStatus` explains the reason.
-            )
-        )
+        if isinstance(column.structural_definition,(QbDimension, QbMultiUnits, QbMultiMeasureDimension)):
+            return True
+        elif (isinstance(column.structural_definition, QbAttribute) and column.structural_definition.get_is_required()):
+            if self.cube.is_pivoted_shape:
+                #If the cube is in pivoted shape, the attribute columns cannot be set to required.
+                _logger.warning(
+                    "Setting the column '%s' to not required as the cube is in pivoted shape.", column.csv_column_title
```
Comment (robons) [inline src/csvcubed/writers/qbwriter.py:549]: Can you alter this message to say something like `Attribute column '%s' was marked as required, but the cube has multiple observed values columns. Attributes in multi-measure pivoted cubes cannot currently be marked as required.`

File: src/csvcubed/writers/qbwriter.py [hunk 2]
```diff
@@ -540,19 +540,19 @@ def _get_primary_key_columns(self) -> List[str]:
...
-                isinstance(column.structural_definition, QbObservationValue)
-                and len(get_observation_status_columns(self.cube)) == 0
-                # We cannot mark an observation value column as `required` if there are `obsStatus` columns defined
-                # since we permit missing observation values where an `obsStatus` explains the reason.
-            )
-        )
+        if isinstance(column.structural_definition,(QbDimension, QbMultiUnits, QbMultiMeasureDimension)):
+            return True
+        elif (isinstance(column.structural_definition, QbAttribute) and column.structural_definition.get_is_required()):
+            if self.cube.is_pivoted_shape:
```
Comment (robons) [inline src/csvcubed/writers/qbwriter.py:546]: It should be possible to have a required attribute in a single measure pivoted data set. We need to check whether it's pivoted **and** it's a multi-obs-val data set in this condtion here.

File: tests/unit/writers/qbwriter/test_qbwriter.py [hunk 3]
```diff
@@ -199,7 +200,10 @@ def test_csv_col_required():
         csv_col = empty_qbwriter._generate_csvw_column_definition(col)
         required = csv_col["required"]
         assert isinstance(required, bool)
-        assert required
+        if isinstance(col.structural_definition, QbAttribute):
+            assert required == False
```
Comment (robons) [inline tests/unit/writers/qbwriter/test_qbwriter.py:204]: This should only happen in a cube with multiple obs val columns. This cube has only one obs-val column. And perhaps more concerningly, this isn't even a pivoted shape cube.

**Review by robons — CHANGES_REQUESTED** [review]

File: tests/unit/writers/qbwriter/test_qbwriter.py (hunk 3 above)
Comment (nimshi89) [reply tests/unit/writers/qbwriter/test_qbwriter.py]: Robsays hi

**Review by nimshi89 — COMMENTED** [review]
I am smart.

**Review by robons — APPROVED** [review]
I approve.

**Pull request merged**